label_weeds.py

```diff
@@ -74,38 +65,83 @@ def make_gray_im(self, color_path, method="ndvi"):
             gray = get_com_im(exg, ndvi)
         else:
             assert False, f"unknown method: {method}"
-        gray = cv2.medianBlur(gray, 5)
+        # gray = cv2.medianBlur(gray, 5)
         # gray = cv2.equalizeHist(gray)
         hist_im = self.make_hist_im(gray)
         return gray, hist_im
 
-    def save_data(self):
-        with open(self.out_json, "w") as outfile:
-            json.dump(self.data, outfile, indent=4)
+    def save_data_voc(self):
+        for image_name, image_data in self.data.items():
+            root_elem = ET.Element("annotation")
+
+            filename_elem = ET.SubElement(root_elem, "filename")
+            filename_elem.text = os.path.basename(image_name)
+
+            bbox_list = image_data.get("bbox_list", [])
+            for bbox in bbox_list:
+                label = bbox[4]
+                if label:
+                    object_elem = self.make_object_element(bbox[:4], label)
+                    root_elem.append(object_elem)
+
+            tree = ET.ElementTree(root_elem)
+            xml_path = os.path.join(self.out_dir, f"{os.path.splitext(os.path.basename(image_name))[0]}.xml")
+            tree.write(xml_path, encoding="utf-8", xml_declaration=True)
```

Comment:
Ten xml jde určitě zapsat nějak lidštěji, tj. ne vše do jedné řádky.. né že bych věděl jak, ale podíval bych se na to.

Reply:
minidom?
https://docs.python.org/3.10/library/xml.dom.minidom.html